futures-util/src/future/future/flatten.rs

```diff
@@ -1,56 +1,84 @@
-use super::chain::Chain;
-use core::fmt;
+use core::fmt::{self, Debug};
 use core::pin::Pin;
 use futures_core::future::{FusedFuture, Future};
 use futures_core::task::{Context, Poll};
-use pin_utils::unsafe_pinned;
+use pin_project::pin_project;
 
-/// Future for the [`flatten`](super::FutureExt::flatten) method.
-#[must_use = "futures do nothing unless you `.await` or poll them"]
-pub struct Flatten<Fut>
+#[pin_project]
+#[derive(Debug)]
+enum InternalFlatten<Fut: Future> {
+    First(#[pin] Fut),
+    Second(#[pin] Fut::Output),
+    Empty,
+}
+
+impl<Fut: Future> InternalFlatten<Fut> {
+    fn new(future: Fut) -> Self {
+        InternalFlatten::First(future)
+    }
+}
+
+impl<Fut> FusedFuture for InternalFlatten<Fut>
     where Fut: Future,
+          Fut::Output: Future,
 {
-    state: Chain<Fut, Fut::Output, ()>,
+    fn is_terminated(&self) -> bool {
+        match self {
+            InternalFlatten::Empty => true,
+            _ => false,
+        }
+    }
 }
 
-impl<Fut> Flatten<Fut>
+impl<Fut> Future for InternalFlatten<Fut>
     where Fut: Future,
           Fut::Output: Future,
 {
-    unsafe_pinned!(state: Chain<Fut, Fut::Output, ()>);
+    type Output = <Fut::Output as Future>::Output;
 
-    pub(super) fn new(future: Fut) -> Flatten<Fut> {
-        Flatten {
-            state: Chain::new(future, ()),
-        }
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        Poll::Ready(loop {
+            match self.as_mut().project() {
+                __InternalFlattenProjection::First(f) => {
```

Comment:
This is an internal type of the generated code which shouldn't be used. You can use the `#[project]` attribute to allow matching on the projected variants.

Reply:
Yeah, I'ved fixed that in the larger PR I have :)